src/wavefront-reporter.js

```diff
@@ -0,0 +1,172 @@
+    enableRuntimeMetrics = false
+  }) {
+    super(registry);
+    this.server = server;
+    this.source = source;
+    this.batchSize = 10000;
+    this.wavefrontClient = new wavefrontSDK.WavefrontDirectClient({
+      server: server,
+      token: token,
+      batchSize: this.batchSize,
+      flushIntervalSeconds: reportingInterval
+    });
+    this.histogramGranularity = new Set();
+    this.flushCurrentHist = true;
+    this.tags = !utils.isEmpty(tags) ? tags : {};
+  }
+
+  /**
+   * Collect metrics from registry and report them to Wavefront.
+   * TODO: With option to include current minute bin of histogram.
+   */
+  report() {
+    // Get all the metrics
+    let allMetrics = this.getMetrics();
+    for (let metrics of [
+      allMetrics.counters,
+      allMetrics.gauges,
+      allMetrics.meters,
+      allMetrics.timers,
+      allMetrics.histograms
+    ]) {
+      if (metrics.length > 0) {
+        for (let m of metrics) {
+          let [metricName, tags] = this._decodeMetric(m);
+          // If Wavefront histogram, client send_distribution
+          if (m instanceof WavefrontHistogram) {
+            let distributions = m.getDistribution();
+            if (this.flushCurrentHist) {
+              distributions.push(m.getCurrentMinuteDistribution());
+            }
+            for (let dist of distributions) {
+              this.wavefrontClient.sendDistribution(
+                metricName,
+                dist.centroids,
+                this.histogramGranularity,
+                dist.timestamp,
+                this.source,
+                tags
+              );
+            }
+            continue;
+          }
+          // If metric, client send_metric
+          for (let f of metricsFunctionMap[m.type]) {
+            const [suffix, value] = f(m);
+            console.log(metricName + suffix + ': ' + value);
+            this.wavefrontClient.sendMetric(
+              metricName + suffix,
+              value,
+              Date.now(),
+              this.source,
+              tags
+            );
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Decode name and tags of encoded metric.
+   * @param metric
+   * @returns {[string, any]}
+   */
+  _decodeMetric(metric) {
+    const [metricName, metricTagsArray] = utils.decodeKey(metric.name);
+    const tagsArray = JSON.parse(metricTagsArray);
+    let metricTags = {};
+    if (tagsArray && tagsArray.length) {
+      tagsArray.forEach(([key, value]) => (metricTags[key] = value));
+    }
+    const tags = Object.assign({}, metricTags, this.tags);
+    return [metricName, tags];
+  }
+
+  /**
+   * Flush the data and close the wavefront client
+   */
+  stop() {
+    this.report();
+    this.wavefrontClient.close();
+  }
+
+  /**
+   * Report distribution using minute granularity.
+   * @returns {WavefrontReporter}
+   */
+  reportMinuteDistribution() {
+    this.histogramGranularity.add(wavefrontSDK.histogramGranularity.MINUTE);
+    return this;
```

Comment:
Why return `this`?

Reply:
see example here: https://github.com/wavefrontHQ/wavefront-nodejs-metrics-sdk/blob/yzheqing/dev/example.js#L15
we might need the reference to the reporter after this function is called to set the granularity

Reply:
hmm this style looks weird to me. If you were able to call this function, it means that you already had a reference to the class instance as well. In the example, you can do the following so that this function doesn't have to return `this`.
```
const wfReporter = new WavefrontReporter({
    server: server,
    token: token,
    source: 'wavefront-javascript-metrics',
    registry: reg,
    tags: { key1: 'val1', key2: 'val2' }
});
wfReporter.reportMinuteDistribution();
wfReporter.start(5000);
```

Reply:
This is something similar to a builder pattern. We need only one statement for creating an instance as well as setting its granularity. Furthermore, you could chain up like `new WavefrontReporter(...).reportMinuteDistribution(). reportHourDistribution().reportDayDistribution()`